CMakeLists.txt

```diff
@@ -1,69 +1,68 @@
-cmake_minimum_required(VERSION 2.8.3)
-project(opw_kinematics)
-
-add_compile_options(-std=c++11 -Wall -Wextra)
-
-find_package(catkin REQUIRED)
+cmake_minimum_required(VERSION 3.5.0)
```

Comment:
Don't we need at least `3.8` here? Afaik, `target_compile_features(${PROJECT_NAME} INTERFACE cxx_std_11)` is only supported since that version.

That would be a problem, as Xenial ships with CMake `3.5`.

Reply:
https://cmake.org/cmake/help/v3.5/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#prop_gbl:CMAKE_CXX_KNOWN_FEATURES gives a list of available features. cxx_std_11 is indeed not included in 3.5

Reply:
Just increased the requirement to 3.8. How should this be managed? I believe this would be only be required if building from source but would not cause issues once it is released as Debian. I could switch it using the example below but I would hate to revert if not required once it is released.

```
set_target_properties(myTarget PROPERTIES
    CXX_STANDARD 11
    CXX_STANDARD_REQUIRED YES)
```

Reply:
Another option is to add `-std=c++11` to the compiler options.

Reply:
> I believe this would only be required if building from source but would not cause issues once it is released as Debian.

Any packaging for Xenial will likely use Xenial's version of CMake, which is `3.5`, so it would now fail to even start building `opw_kinematics` as the requirement is now `3.8`.

This is regardless of whether it'd be released using the ROS buildfarm or some other way (assuming you don't want to build on a system where you've replaced the system version of CMake with a newer version of course).

The `set_target_properties(..)` you propose would seem to be a way to achieve a similar result (it's not entirely the same, but probably good enough), although I'm not sure whether target properties are automatically exported.

Reply:
Another option would be to use some of the features already available in `3.5`, ie `cxx_range_for`. Those are then supposed to set `c++11` if required. I don't really like that option however, it's too indirect for my taste

also @Jmeyer1292 if you merge #19 this will all be tested automatically ;)

Reply:
>This is regardless of whether it'd be released using the ROS buildfarm or some other way (assuming you don't want to build on a system where you've replaced the system version of CMake with a newer version of course).

What would be the preferred approach to support new features going forward for pure cmake repositories. I would prefer to not be tied to a specific version of ROS. Is it possible to version releases in the ROS build farm. For example clang, the default release for a distribution is under clang but you can get newer version clang-3.8, etc.

Reply:
It's not about ROS. It's the OS. Xenial != Kinetic.

Reply:
I understand that `Xenial != Kinetic` does not actually have to be the case but if we are limiting capabilities to a minimum version of a library for an OS than they are essentially `kinetic >= Xenial` and it can only leverage system dependencies available versions on Xenial. More specific to ROS build farm. The debian creation process allows you to create a debian that requires build requirements not natively on the OS by leveraging before scripts. It would be possible to create a before script that downloads the correct version of CMake (build dependencies) to build the debian which should not impact other libraries using the released version. Not sure if this is possible through the ROS build farm and it would make it painful for anyone developing on kinetic because they would have to install a new version of CMake to build the package, which is why I was asking about version with ROS build farm. We could have a standard version for the library for a give OS but also provide newer versions for those who would like to use newer capabilities that have breaking changes.

Reply:
The main thing I am trying to understand is the best way to support these types of repositories. In ROS specific repositories we have branches specific to the ROS distribution. Is this the approach that should be used for these types of repositories or should we leverage release version branches with a master branch where all new features would go. Then identify release branches with supported OS and ROS distro. Looking at PCL and VTK they just have a master branch and leverage tags for the releases.

Reply:
Ah, I had not understood your comment as one asking about a branching/release model.

Using a single `master` branch with some additional branches for specific platforms would probably work here. Depending on the branching style this could also be done with patches during the build process but I don't necessarily like that very much.

Having multiple `master` branches (ie: an actual `master` and then master per release target/OS/platform) could be considered. `master` contains cutting-edge, with the others tracking it and applying the fixes needed to use the product on different platforms.

Other workflows exist, but would need a completely different branching model.

Reply:
So I understand correctly we should create a branch for xenial, bionic, etc. Keep the master that contains the cutting edge. In the case of this cmake version, keep the 3.8 requirement then create the xenial and bionic branch and then create a PR that makes the change to xenial branch that reduces the cmake version and adds the alternative method for setting the c++ version?

Reply:
That would be one way to approach this, yes.

There are most likely much more formal development and release workflows that could deal with this sort of situation though. I haven't looked anything up. This would seem to allow us to have a `master` branch for cutting edge and use the per-target-OS branch to resolve issues with packaging and other meta-topics.

We could also go for a separate repository entirely for versioning packaging artefacts (scripts, patches, manifests). That would be closer to how it should be done, but also introduces quite a bit of overhead.

This may be a good compromise.

Reply:
Ok so are @gavanderhoorn  good with merging this. Then I will create a xenial branch and the submit a PR that patches it so it will build.

Reply:
Is that really worth the trouble? It should work with cmake 3.5 features just fine right?

Reply:
> Is that really worth the trouble? It should work with cmake 3.5 features just fine right?

The target compiler feature is preferred because it is better supported for cross platform. Given ROS-2 is cross platform and we are working to leverage this in ROS2 on linux and windows soon this would simplify things. Also it supports interface targets which is nice. [This](https://cliutils.gitlab.io/modern-cmake/chapters/features/cpp11.html) provides good insight into the advantages.

Reply:
`set_target_properties(..)` also results in a cross platform `CMakeLists.txt`. It's just that `target_compile_features(..)` uses a feature/capability based detection system, which makes it a little nicer.

What would be a real problem would be to keep the `add_compile_options(-std=c++11 ...)` as that is incompatible with anything but GCC and CLang.

Reply:
(more humorous side-comment first: actually `-std=c++11` does "work" on msvc ... https://github.com/protocolbuffers/protobuf/issues/4570 it will just warn about an unused option but msvc defaults to MS-C++ anyway (it supports whatever was standardized/implemented at release time, so it's a subset of all standards, none being complete AFAIK) )

`target_compile_features` could also be used with the (more fine-grained) features available in cmake 3.5 (`auto`, `using`, ... are `c++11` features currently used ). Then there would be no need for a separate branch yet/ever?

Reply:
AFAIK, the set_target_properties have a private scope and does not support interface target meaning this requirement of c++11 would not get passed to any target using this interface target. This requires the individual using this interface target to also add the set_target_property for c++11.

Reply:
Also, I don't understand why we would not want the master branch to move forward and why would we hold back the progression for an LTS that is in the maintenance stage of it life cycle. Currently CMake is on version 3.14 and bionic is released with 3.10. I personally think the master should stay as up to date with latest capabilities and the older released version be burdened with requiring patches.

Reply:
> Also, I don't understand why we would not want the master branch to move forward and why would we hold back the progression for an LTS that is in the maintenance stage of it life cycle.

Because branching always introduces maintenance overhead. If branching (for these kind of things) can be avoided, that is definitely beneficial.

Reply:
> Because branching always introduces maintenance overhead. If branching (for these kind of things) can be avoided, that is definitely beneficial.

I complete agree, that this causes maintenance overhead and it is beneficial to the maintainer. I personally would vote to keep the master as updated as possible and take on the extra maintenance to keep things advancing and when older version are no longer maintainable mark it as no longer supported and direct them to the newer version if they are needing particular feature.